### PR TITLE
Feature/alb augmix

### PR DESCRIPTION
**Pull request by sirn524**
## 実施したこと

* Albumentationsにaug_mixの分岐を実装
* segmentations_aug.pyのリファクタリングと一部Kerasモジュールの削除
* yaml上でAugmentationの実行する種類を選べるようになる（と言っても選択肢は今の所Albumentationsのみ）

## やらないこと

* Keras(ImageDataGenerator)によるAugmentation分岐の実装
* Albumentations以外のAugmentationModuleの実装

## できるようになること（ユーザ目線）

* Augmixによる複雑なaugmentation
* より多くのAugmentationモジュールの受け入れ準備
* yamlにてaugmix : yesでaugmixへの分岐を実行できるようになった

## <del>できなくなること（ユーザ目線）<del> 

<del> * yamlのaugmentation_statを明記しないと、augmentationが実行されないようになった</del> 

## 動作確認

* DL2にてexampleでの動作確認
* GitHub Actionsでの動作確認

## その他

* yamlからの操作が少し複雑になりました。augmentation_stat, augmixの入力が必要になります。
* augmixモジュールの導入による実行効果の実験はしていません。精度が下がる場合にはaugmix: no(false)をdefault, exampleにも反映します。

### CODE REVIEW

**Comment by sirn524** [comment]
Conflictの解消後、train_paramsのfeatureがマージされたため、train_paramsにaugmix parameter2つを追加
追加要項
・augmentation_stat --- augmentationの種類が増えるに連れて、yamlでの操作が可能となる
・augmix --- true, falseでmixをするかどうかを決定する。

**Comment by sirn524** [comment]
train paramsのパラメータ追加が良かったのかどうかは応相談と思われる。
augmix true-- booleanはいるべきか？まぁ、cosine_decayもbool値あるしOK？

File: farmer/ncc/augmentation/segmentation_aug.py [hunk 1]
```diff
@@ -4,15 +4,31 @@
...
+        if augmix:
+            augmented = augment_and_mix(augmented[image], mean, std,)
         return augmented['image'], augmented["mask"]
 
-    else:
+    elif "imagegenerator" in aug_stat:
+    # will be modified with keras_image_generator
+        if augmix:
+            augmented = augment_and_mix(input_image, mean, std)
+            label = label
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/segmentation_aug.py:31]: augmixがTrue or Flaseさえあれば十分だと思ったけど、
`aug_stat`が必要な理由って何？

File: farmer/domain/model/train_params_model.py [hunk 2]
```diff
@@ -18,4 +18,7 @@ class TrainParams:
     class_weights: Dict[int, float] = field(default_factory=dict)
     learning_rate: float = None
     optimizer: str = None
+    augmentation_stat: str = None
     augmentation: List[str] = field(default_factory=list)
+    augmix: bool = False
+
```
Comment (aiorhiroki) [inline farmer/domain/model/train_params_model.py:24]: 最終行

Comment (aiorhiroki) [inline farmer/domain/model/train_params_model.py:24]: 最終行がないよ

File: farmer/ncc/augmentation/segmentation_aug.py [hunk 3]
```diff
@@ -4,15 +4,31 @@
 import albumentations
 
 
-def segmentation_alb(input_image, label, mean, std, augmentation_dict):
+def segmentation_alb(
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/segmentation_aug.py:7]: 関数名は`segmentation_aug`の方がよさそう
alb以外も使う統合された関数になったから。

File: farmer/ncc/augmentation/segmentation_aug.py (hunk 1 above)
Comment (sirn524) [reply farmer/ncc/augmentation/segmentation_aug.py:31]: 山田さんが起票した、オリジナルのaugmixにkeras ImageDataGeneratorの分岐があったので、そこの追加がしたくて作りました。(スコープからは外れていましたので、Generator自体の実装はしていないですが）このfeatureだと必要ないということであれば、消してもいいかもです

File: farmer/ncc/augmentation/segmentation_aug.py (hunk 3 above)
Comment (sirn524) [reply farmer/ncc/augmentation/segmentation_aug.py:7]: そうですね、augに書き換えします！

File: farmer/ncc/augmentation/segmentation_aug.py (hunk 1 above)
Comment (aiorhiroki) [reply farmer/ncc/augmentation/segmentation_aug.py:31]: このfeatureでは一旦消して良さそう。また追加された時に復活する感じで。

あと、`label = label`はいらない 👼

File: farmer/ncc/augmentation/segmentation_aug.py [hunk 4]
```diff
@@ -4,15 +4,31 @@
...
 
-    if len(transforms) > 0:
+    if aug_stat is None:
+        return input_image, label
+
+    elif "albumentation" in aug_stat:
         aug = albumentations.Compose(transforms, p=1)
         augmented = aug(image=input_image, mask=label)
+        if augmix:
+            augmented = augment_and_mix(augmented[image], mean, std,)
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/segmentation_aug.py:24]: **std,)** 最後にごみある

File: farmer/ncc/augmentation/segmentation_aug.py (hunk 1 above)
Comment (sirn524) [reply farmer/ncc/augmentation/segmentation_aug.py:31]: 了解です

**Comment by yumion** [comment]
> 山田さんが起票した、オリジナルのaugmixにkeras ImageDataGeneratorの分岐があったので、そこの追加がしたくて作りました。

これ、ImageDataGeneratorを使っていたときにAugMixを入れただけなので、ImageDataGeneratorは必要ないですよ。
あと、 `augment_and_mix` は入力画像だけにしかaugmentationをしないので、labelもaugmentationした方を使っていただけると嬉しいです（そっちはalbumentationsで実装もしてますし）

**Comment by sirn524** [comment]
まず松崎さんからいただいた指摘部分反映後、label(mask)をaugmixするように組み込みを分解して考えることにします。
ImageDataGeneratorで目新しいAugmentationとかないですかね。。。random erasingとかも組み込みたいので、別featureで追加（ImageDataGenerator以外でも）検討します

**Comment by yumion** [comment]
> ImageDataGeneratorで目新しいAugmentationとかないですかね

ないと思います…

**Comment by sirn524** [comment]
指摘いただいたパートはresolveしましたー

**Comment by sirn524** [comment]
label,maskの両方がaugment and mixに入るようにしてみた

File: farmer/ncc/augmentation/segmentation_aug.py [hunk 5]
```diff
@@ -4,16 +4,21 @@
...
+def segmentation_aug(
+        input_image,
+        label, mean,
+        std,
+        augmentation_dict,
+        augmix,
+):
     transforms = get_aug(augmentation_dict)
 
-    if len(transforms) > 0:
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/segmentation_aug.py:10]: `if len(transforms) > 0:`
これなくて良いの？

File: farmer/ncc/augmentation/segmentation_aug.py [hunk 6]
```diff
@@ -4,16 +4,21 @@
...
-        augmented = aug(image=input_image, mask=label)
-        return augmented['image'], augmented["mask"]
-
-    else:
-        return input_image, label
+    aug = albumentations.Compose(transforms, p=1)
+    augmented = aug(image=input_image, mask=label)
+    if augmix:
+        augmented['image'] = augment_and_mix(augmented['image'], mean, std)
+        augmented['mask'] = augment_and_mix(augmented['mask'], mean, std)
```
Comment (yumion) [inline farmer/ncc/augmentation/segmentation_aug.py:20]: これ、imageとmask一致しないですよ
ランダムにaugmentation選ぶので、別で実行すると別のaugmentationがmixされます

File: farmer/ncc/augmentation/segmentation_aug.py (hunk 5 above)
Comment (sirn524) [reply farmer/ncc/augmentation/segmentation_aug.py:10]: 消しちゃってました、、、絶対必要です

File: farmer/ncc/augmentation/segmentation_aug.py (hunk 6 above)
Comment (yumion) [reply farmer/ncc/augmentation/segmentation_aug.py:20]: ここを参考にしてください
https://github.com/Project-zero-one/Hypogastric_nerve/blob/b5ed957e70ba65f91273d9101c3212f0048f1fbc/src/augmentations.py#L217

File: farmer/ncc/augmentation/segmentation_aug.py [hunk 7]
```diff
@@ -4,17 +4,23 @@
...
         aug = albumentations.Compose(transforms, p=1)
         augmented = aug(image=input_image, mask=label)
+        if augmix:
+            augmented['image'] = augment_and_mix(augmented['image'], mean, std)
+            augmented['mask'] = augment_and_mix(augmented['mask'], mean, std)
         return augmented['image'], augmented["mask"]
 
-    else:
-        return input_image, label
-
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/segmentation_aug.py:17]: この3行も消しちゃまずそう

File: farmer/ncc/augmentation/augment_and_mix.py [hunk 8]
```diff
@@ -32,7 +32,52 @@ def apply_op(image, op, severity):
     return np.asarray(pil_img) / 255.
 
 
-def augment_and_mix(image, mean, std, severity=3, width=3, depth=-1, alpha=1., augall=False):
+def dual_augment_and_mix(image, mask, transforms, mean, std, width=3, depth=-1, alpha=1.):
+    """Perform AugMix augmentations and compute mixture.
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/augment_and_mix.py:36]: segmentationにしか、augmixできなくなるけど、classificationでもやるよね。
元々のaugment_and_mix消す必要ないと思う。

File: farmer/ncc/augmentation/segmentation_aug.py [hunk 9]
```diff
@@ -1,15 +1,27 @@
...
 import numpy as np
-from .augment_and_mix import augment_and_mix
+from .augment_and_mix import dual_augment_and_mix
 import albumentations
 
 
-def segmentation_alb(input_image, label, mean, std, augmentation_dict):
+def segmentation_aug(
+        input_image,
+        label, mean,
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/segmentation_aug.py:9]: ここ改行しよう

File: farmer/ncc/augmentation/segmentation_aug.py [hunk 10]
```diff
@@ -1,15 +1,27 @@
...
     transforms = get_aug(augmentation_dict)
 
     if len(transforms) > 0:
         aug = albumentations.Compose(transforms, p=1)
         augmented = aug(image=input_image, mask=label)
+        if augmix:
+            augmented = dual_augment_and_mix(
+                augmented['image'],
+                augmented['mask'],
+                mean, std
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/segmentation_aug.py:23]: このdual_augment_and_mixにyamlで指定したtransformsを入れるんじゃないの？

File: example/yamls/segmentation.yaml [hunk 11]
```diff
@@ -49,7 +49,7 @@ train_params:
             MotionBlur:
                 p: 1  # percentage of function inside 'OneOf'
             p: 1  # percetage of 'OneOf'
-
+    augmix: yes
```
Comment (aiorhiroki) [inline example/yamls/segmentation.yaml:52]: できるなら、augmentationの中に、augmixがあった方が見え方が綺麗だよね。
難しいかな？

**Review by aiorhiroki — CHANGES_REQUESTED** [review]
classificationの対応と、yamlのtransformsを使うaugmixをすること。

File: farmer/ncc/augmentation/augment_and_mix.py (hunk 8 above)
Comment (sirn524) [reply farmer/ncc/augmentation/augment_and_mix.py:36]: 改行の関係でaugment_and_mixが下に行っているため、消したように見えているだけっぽいです。
ファイル変更なしでresolveしてもよろしいですか？

Comment (sirn524) [reply farmer/ncc/augmentation/augment_and_mix.py:36]: 見え方悪いですが、オリジナルのaugment_and_mixは残っています。行数の関係で場所を入れ替えたので消えたように記述されているのかと思います。

File: farmer/ncc/augmentation/segmentation_aug.py (hunk 9 above)
Comment (sirn524) [reply farmer/ncc/augmentation/segmentation_aug.py:9]: 承知しました、確かに文字数ではなく要素で言ったら独立ですね、、、失礼しました

File: farmer/ncc/augmentation/augment_and_mix.py (hunk 8 above)
Comment (aiorhiroki) [reply farmer/ncc/augmentation/augment_and_mix.py:36]: あ、ほんとだね
ClassificationDatasetにaugmentationのコード書いて欲しいなー

Comment (sirn524) [reply farmer/ncc/augmentation/augment_and_mix.py:36]: わかりました、やってみます

File: example/yamls/segmentation.yaml (hunk 11 above)
Comment (aiorhiroki) [reply example/yamls/segmentation.yaml:52]: パッと見、難しそうだね。このままでいいか

File: farmer/ncc/augmentation/segmentation_aug.py (hunk 6 above)
Comment (sirn524) [reply farmer/ncc/augmentation/segmentation_aug.py]: dual_augmentの変更により解消済み

File: farmer/ncc/augmentation/segmentation_aug.py (hunk 10 above)
Comment (sirn524) [reply farmer/ncc/augmentation/segmentation_aug.py:25]: transforms抜けてました。なぜ気づかなかったか不思議なくらいです。すみません

File: farmer/ncc/augmentation/augment_and_mix.py (hunk 8 above)
Comment (sirn524) [reply farmer/ncc/augmentation/augment_and_mix.py:36]: classification 側はaugment_and_mixのみで対応するように書き換えました。

File: farmer/ncc/generators/dataset.py [hunk 12]
```diff
@@ -107,6 +110,15 @@ def __getitem__(self, i):
         input_image = self.image_util.normalization(input_image)
         label = self.image_util.cast_to_onehot(label)
 
+        # apply augmentations
+        if self.augmentation and len(self.augmentation) > 0:
+            input_image, label = classification_aug(
```
Comment (aiorhiroki) [inline farmer/ncc/generators/dataset.py:115]: classification_augが定義されてないよ

**Review by aiorhiroki — CHANGES_REQUESTED** [review]
segmentation_aug.pyの中にclassification_augを書くのは違うから、
classification_aug.pyを作るか、albumentation_aug.pyで二つをまとめるか、どちらかが必要

File: farmer/ncc/augmentation/classification_aug.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        mean,
+        std,
+        augmentation_dict,
+        augmix,
+):
+    transforms = get_aug(augmentation_dict)
+
+    if len(transforms) > 0:
+        aug = albumentations.Compose(transforms, p=1)
+        augmented = aug(image=input_image, mask=label)
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/classification_aug.py:19]: classificationの時は、mask,labelどちらも必要ないね

File: farmer/ncc/augmentation/classification_aug.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+                transforms,
+                mean,
+                std
+            )
+        return augmented['image'], augmented["mask"]
+
+    else:
+        return input_image, label
+
+def get_aug(augmentation_dict):
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/classification_aug.py:33]: 同じ関数が出てきてるのは美しくないから、ここは共通化しておきたいな。
utils.pyとか作れば良さそう

**Review by aiorhiroki — COMMENTED** [review]
コンフリクト解消もお願いします...

**Comment by sirn524** [comment]
一旦conflictのresolveとclassificationのいらないパートを消す等のreviewいただいたパートは全部修正しました。
Classificationの挙動についてもう少し調べたいので、review指摘頂きつつでも良いですか？

File: farmer/ncc/augmentation/augment_and_mix.py [hunk 15]
```diff
@@ -50,12 +95,15 @@ def augment_and_mix(image, mean, std, severity=3, width=3, depth=-1, alpha=1., a
         np.random.dirichlet([alpha] * width))
     m = np.float32(np.random.beta(alpha, alpha))
 
-    mix = np.zeros_like(image)
+    image_mix = np.zeros_like(image)
```
Comment (aiorhiroki) [inline farmer/ncc/augmentation/augment_and_mix.py:98]: これ`mix`を`image_mix`に変えて問題ない？

**Review by aiorhiroki — COMMENTED** [review]
差分が多くなってしまっているけど、たぶん良さそう！

File: farmer/ncc/augmentation/augment_and_mix.py (hunk 15 above)
Comment (sirn524) [reply farmer/ncc/augmentation/augment_and_mix.py:98]: def augment_and_mixは、classificationしか使わないので、影響はないと思います。
for文の中のmix が変わってなかったので一部修正します。
